### PR TITLE
Store self-signed key/cert in secret so next time CA does not regenerate

### PR DESCRIPTION
**Pull request by wattli**
… write it to secret for persistency

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/pki/ca/ca.go [hunk 1]
```diff
@@ -80,6 +105,47 @@ func NewSelfSignedIstioCA(caCertTTL, certTTL time.Duration, org string) (*IstioC
 // NewIstioCA returns a new IstioCA instance.
 func NewIstioCA(opts *IstioCAOptions) (*IstioCA, error) {
 	ca := &IstioCA{certTTL: opts.CertTTL}
+	// If the signing key/cert or root cert is empty, we should create a self-signed key/cert pair,
+	// and write it to the secret for persistent purpose.
+	// TODO(wattli): get rid of the NewSelfSignedIstioCA() after 0.2.
+	// K8s secret has a bug that you can't specify empty content, so we put a small content in it and
```
Comment (lookuptable) [inline pkg/pki/ca/ca.go]: Can we link to the bug/issue here? 

Who is "we" when you say "we put a small content in it"?

File: pkg/pki/ca/ca.go [hunk 2]
```diff
@@ -22,10 +22,29 @@ import (
...
+	IstioSecretType = "istio.io/key-and-cert"
+
+	// The ID/name for the certificate chain file.
+	CACertChainID = "ca-cert.pem"
+	// The ID/name for the private key file.
+	CAPrivateKeyID = "ca-key.pem"
+	// The ID/name for the CA root certificate file.
+	RootCertID = "root-cert.pem"
+
+	CAServiceAccount = "istio-ca-creds"
```
Comment (lookuptable) [inline pkg/pki/ca/ca.go]: We can't assume the service account is a const, can we?

File: pkg/pki/ca/ca.go [hunk 3]
```diff
@@ -80,6 +105,47 @@ func NewSelfSignedIstioCA(caCertTTL, certTTL time.Duration, org string) (*IstioC
...
+		// Rewrite the key/cert back to secret so they will be persistent when CA restarts.
+		secret := &apiv1.Secret{
+			Data: map[string][]byte{
+				CACertChainID:  pemCert,
+				CAPrivateKeyID: pemKey,
+				RootCertID:     pemCert,
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Annotations: map[string]string{serviceAccountNameAnnotationKey: CAServiceAccount},
+				Name:        SecretNamePrefix + CAServiceAccount,
```
Comment (lookuptable) [inline pkg/pki/ca/ca.go]: So if an admin specifies an empty secret called `ca-signing-secret` and mounts it into the CA pod, the CA will in turn write the self-signed key/cert into a totally different secret (e.g. `istio-ca.istio-ca-creds`). This would be quite confusing, right? And it won't help in case the ca pod is recreated.

File: pkg/pki/ca/ca.go [hunk 4]
```diff
@@ -22,10 +22,29 @@ import (
 	"errors"
 	"time"
 
+	"github.com/golang/glog"
 	"istio.io/auth/pkg/pki"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (lookuptable) [inline pkg/pki/ca/ca.go]: I'd be careful in bringing dependency on K8S apis here. Previously CA is CA and all interactions with K8S are delegated to the `controller`.

File: pkg/pki/ca/ca.go (hunk 3 above)
Comment (wattli) [reply pkg/pki/ca/ca.go]: Good point, an easy way is to document it clearly not to mess up with the default CA config. It might worth to make it a flag going forward.

File: pkg/pki/ca/ca.go (hunk 2 above)
Comment (wattli) [reply pkg/pki/ca/ca.go]: It's actually not a service account, it's the secret name. Do we actually need this annotation?

File: pkg/pki/ca/ca.go (hunk 3 above)
Comment (wattli) [reply pkg/pki/ca/ca.go]: Good point. The secret (with dummy content) needs to be there before deploying istio-ca, and we will have a yaml file to create that secret (in front of istio-ca yaml). 

An easy way is to document it clearly not to mess up the default CA secret name (and I doubt people will actually do that). 

Of course this can be done via flag, but we will need admin to specify secret name at two places: where we do volume mount and CA input params. Personally I prefer not to introduce another flag to CA.

**Comment by wattli** [comment]
@lookuptable  PTAL, thanks

File: cmd/istio_ca/main.go [hunk 5]
```diff
@@ -172,14 +174,17 @@ func createCA() ca.CertificateAuthority {
 	caOpts := &ca.IstioCAOptions{
 		CertChainBytes:   certChainBytes,
 		CertTTL:          opts.certTTL,
+		Core:             core,
+		Namespace:        opts.namespace,
```
Comment (lookuptable) [inline cmd/istio_ca/main.go]: What happens when namespace is unspecified (i.e., in cluster-wide deployment)?

File: pkg/pki/ca/ca.go (hunk 1 above)
Comment (lookuptable) [reply pkg/pki/ca/ca.go]: The comment seems to be the same as before.

File: pkg/pki/ca/ca.go (hunk 3 above)
Comment (lookuptable) [reply pkg/pki/ca/ca.go]: hmm I feel it's quite a restriction to force the secret to be named as `isito-ca.istio-ca-creds`. 

IMO the idea workflow would be:

1. Admin creates a secret with an arbitrarily chosen name
1. The admin mounts the secret into CA at a well-defined dir
1. CA examines the dir, and write key/cert if the dir is empty
1. The content that CA writes is auto fed back to apiserver (we might need to verify if this is possible)

If step 4 is not possible, I'd vote to explicitly passing in a secret name.

File: pkg/pki/ca/ca.go [hunk 6]
```diff
@@ -80,6 +107,47 @@ func NewSelfSignedIstioCA(caCertTTL, certTTL time.Duration, org string) (*IstioC
...
+		secret := &apiv1.Secret{
+			Data: map[string][]byte{
+				CACertChainID:  pemCert,
+				CAPrivateKeyID: pemKey,
+				RootCertID:     pemCert,
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Annotations: map[string]string{serviceAccountNameAnnotationKey: CAServiceAccount},
+				Name:        SecretNamePrefix + CAServiceAccount,
+				Namespace:   opts.Namespace,
```
Comment (lookuptable) [inline pkg/pki/ca/ca.go]: This will like cause problem as Namespace could be empty.

Comment (wattli) [reply pkg/pki/ca/ca.go]: Fixed.

File: cmd/istio_ca/main.go (hunk 5 above)
Comment (wattli) [reply cmd/istio_ca/main.go]: Fixed.

**Comment by wattli** [comment]
@lookuptable , please take a look, need to get this in ASAP. I will ask IBMer to review after your approval.

**Comment by lookuptable** [comment]
I'll try to get to this tonight. Feel free to bring a second reviewer meanwhile to speed it up.

File: cmd/istio_ca/main.go [hunk 7]
```diff
@@ -49,8 +50,9 @@ type cliOptions struct {
 	signingKeyFile  string
 	rootCertFile    string
 
-	namespace      string
-	kubeConfigFile string
+	namespace          string
+	targetAllNamespace bool
```
Comment (lookuptable) [inline cmd/istio_ca/main.go]: Nit: AllNamespaces

File: cmd/istio_ca/main.go [hunk 8]
```diff
@@ -82,19 +84,22 @@ func init() {
...
 
 	flags.BoolVar(&opts.selfSignedCA, "self-signed-ca", false,
 		"Indicates whether to use auto-generated self-signed CA certificate. "+
 			"When set to true, the '--signing-cert' and '--signing-key' options are ignored.")
+
 	flags.StringVar(&opts.selfSignedCAOrg, "self-signed-ca-org", "k8s.cluster.local",
 		fmt.Sprintf("The issuer organization used in self-signed CA certificate (default to %s)",
 			selfSignedCAOrgDefault))
 
+	// TODO(wattli): change it to IntVar to make it configurable in yaml.
```
Comment (lookuptable) [inline cmd/istio_ca/main.go]: Sorry I don't agree with this todo. Right now you can already specify a string that can be parsed as a duration object (e.g. 5m, 10h)

File: cmd/istio_ca/main.go [hunk 9]
```diff
@@ -121,13 +126,21 @@ func runCA() {
...
 
 	verifyCommandLineOptions()
 
-	ca := createCA()
 	cs := createClientset()
-	sc := controller.NewSecretController(ca, cs.CoreV1(), opts.namespace)
+	ca := createCA(cs.CoreV1())
+	controllerNs := opts.namespace
+	if opts.targetAllNamespace {
+		controllerNs = ""
```
Comment (lookuptable) [inline cmd/istio_ca/main.go]: Please add comment here

File: pkg/pki/ca/ca.go [hunk 10]
```diff
@@ -22,15 +22,32 @@ import (
...
 
+	"github.com/golang/glog"
 	"istio.io/auth/pkg/pki"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	apiv1 "k8s.io/client-go/pkg/api/v1"
 )
 
 const (
+	// IstioSecretType is the Istio secret annotation type.
```
Comment (lookuptable) [inline pkg/pki/ca/ca.go]: Not: all constants here can be unexported.

File: pkg/pki/ca/ca.go [hunk 11]
```diff
@@ -22,15 +22,32 @@ import (
...
+
+	// CASecret stores the key/cert of self-signed CA for persistency purpose.
+	CASecret = "istio-ca-secret"
+
 	// The size of a private key for a self-signed Istio CA.
 	caKeySize = 2048
 )
 
-// CertificateAuthority contains methods to be supported by a CA.
+// CertificateAuthority contains methods to bfe supported by a CA.
```
Comment (lookuptable) [inline pkg/pki/ca/ca.go]: Please revert this change.

File: pkg/pki/ca/ca.go [hunk 12]
```diff
@@ -40,6 +57,7 @@ type CertificateAuthority interface {
 type IstioCAOptions struct {
 	CertChainBytes   []byte
 	CertTTL          time.Duration
+	Namespace        string
```
Comment (lookuptable) [inline pkg/pki/ca/ca.go]: Guess you don't need this any more.

File: pkg/pki/ca/ca.go [hunk 13]
```diff
@@ -56,24 +74,57 @@ type IstioCA struct {
...
 	opts := &IstioCAOptions{
-		CertTTL:          certTTL,
-		SigningCertBytes: pemCert,
-		SigningKeyBytes:  pemKey,
-		RootCertBytes:    pemCert,
+		CertTTL:   certTTL,
+		Namespace: namespace,
+	}
+	if err != nil {
+		glog.Warningf("Failed to get secret (error: %s), will create one", err)
```
Comment (lookuptable) [inline pkg/pki/ca/ca.go]: Nit: this is a legit case so probably don't log it as warning?

File: pkg/pki/ca/ca.go [hunk 14]
```diff
@@ -56,24 +74,57 @@ type IstioCA struct {
...
+				Namespace: namespace,
+			},
+			Type: IstioSecretType,
+		}
+		_, err := core.Secrets(opts.Namespace).Create(secret)
+		if err != nil {
+			glog.Errorf("Failed to create secret (error: %s)", err)
+			return nil, err
+		}
+	} else {
```
Comment (lookuptable) [inline pkg/pki/ca/ca.go]: add a todo to verify secret content?

File: pkg/pki/ca/ca.go [hunk 15]
```diff
@@ -56,24 +74,57 @@ type IstioCA struct {
...
+		opts.SigningCertBytes = pemCert
+		opts.SigningKeyBytes = pemKey
+		opts.RootCertBytes = pemCert
+
+		// Rewrite the key/cert back to secret so they will be persistent when CA restarts.
+		secret := &apiv1.Secret{
+			Data: map[string][]byte{
+				CACertChainID:  pemCert,
+				CAPrivateKeyID: pemKey,
+				RootCertID:     pemCert,
```
Comment (lookuptable) [inline pkg/pki/ca/ca.go]: Nit: you don't need to save root cert into the secret as it's always identical to the ca cert.

File: pkg/pki/ca/ca.go [hunk 16]
```diff
@@ -56,24 +74,57 @@ type IstioCA struct {
...
-		Org:          org,
-		IsCA:         true,
-		IsSelfSigned: true,
-		RSAKeySize:   caKeySize,
-	}
-	pemCert, pemKey := GenCert(options)
+func NewSelfSignedIstioCA(caCertTTL, certTTL time.Duration, org string, namespace string,
+	core corev1.CoreV1Interface) (*IstioCA, error) {
 
+	caSecret, err := core.Secrets(namespace).Get(CASecret, metav1.GetOptions{})
```
Comment (lookuptable) [inline pkg/pki/ca/ca.go]: I feel it deserves some detailed comment about what is happening here.

File: cmd/istio_ca/main.go (hunk 7 above)
Comment (wattli) [reply cmd/istio_ca/main.go]: done

File: cmd/istio_ca/main.go (hunk 8 above)
Comment (wattli) [reply cmd/istio_ca/main.go]: I see, reverted.

File: cmd/istio_ca/main.go (hunk 9 above)
Comment (wattli) [reply cmd/istio_ca/main.go]: done

File: pkg/pki/ca/ca.go (hunk 11 above)
Comment (wattli) [reply pkg/pki/ca/ca.go]: reverted.

File: pkg/pki/ca/ca.go (hunk 10 above)
Comment (wattli) [reply pkg/pki/ca/ca.go]: They are used in ca_test.go

Comment (lookuptable) [reply pkg/pki/ca/ca.go]: That's in the sane package, right? Still don't need to export them.

Comment (lookuptable) [reply pkg/pki/ca/ca.go]: Same

Comment (wattli) [reply pkg/pki/ca/ca.go]: I see, just followed secret.go example, looks it used somewhere else.

File: pkg/pki/ca/ca.go (hunk 16 above)
Comment (wattli) [reply pkg/pki/ca/ca.go]: done

File: pkg/pki/ca/ca.go (hunk 14 above)
Comment (wattli) [reply pkg/pki/ca/ca.go]: Done

**Comment by wattli** [comment]
@rshriram @ZackButcher , please take a look, this is not a breaking change. It would be great for this to make in 0.2 since customers do not need to manually generate some key/cert for our Istio CA any more. Otherwise it would be a bad user experience for customers.

File: cmd/istio_ca/main.go [hunk 17]
```diff
@@ -82,12 +86,14 @@ func init() {
 	flags.StringVar(&opts.signingKeyFile, "signing-key", "", "Specifies path to the CA signing key file")
 	flags.StringVar(&opts.rootCertFile, "root-cert", "", "Specifies path to the root certificate file")
 
-	flags.StringVar(&opts.namespace, "namespace", "",
-		"Select a namespace for the CA to listen to. If unspecified, Istio CA tries to use the ${"+namespaceKey+"} "+
-			"environment variable. If neither is set, Istio CA listens to all namespaces.")
+	flags.StringVar(&opts.namespace, "namespace", "", "The namesapce that CA is running at.")
+
```
Comment (rshriram) [inline cmd/istio_ca/main.go]: Doesn't this break e2e ?

Comment (rshriram) [reply cmd/istio_ca/main.go]: Shouldn't this default to istio-system ?

File: cmd/istio_ca/main.go [hunk 18]
```diff
@@ -121,13 +127,22 @@ func runCA() {
 		if value, exists := os.LookupEnv(namespaceKey); exists {
 			opts.namespace = value
 		}
+
+		if opts.namespace == "" {
+			glog.Fatal("Fail to get the namespace of this CA")
+		}
```
Comment (andraxylia) [inline cmd/istio_ca/main.go]: This will not work with the current templates in istio/istio, because for cluster-wide, namespace is passed as "".
It is too late to make this change in 0.2

File: cmd/istio_ca/main.go (hunk 17 above)
Comment (wattli) [reply cmd/istio_ca/main.go]: Actually we don't have to, since it will check the env variable about the namespace (see line 127).

File: cmd/istio_ca/main.go (hunk 18 above)
Comment (andraxylia) [reply cmd/istio_ca/main.go]: Suggestion is to make the namespace flags changes in 0.3, since it needs to cascade all the way to pilot and istio, and requires change in the yaml files.

The actual certificate changes lgtm.

**Comment by wattli** [comment]
@costinm @andraxylia , please take a look, thanks.

File: cmd/istio_ca/main.go [hunk 19]
```diff
@@ -85,6 +89,8 @@ func init() {
 	flags.StringVar(&opts.namespace, "namespace", "",
 		"Select a namespace for the CA to listen to. If unspecified, Istio CA tries to use the ${"+namespaceKey+"} "+
 			"environment variable. If neither is set, Istio CA listens to all namespaces.")
+	flags.StringVar(&opts.caNamespace, "ca-namespace", "istio-system", "Specify whether the current CA is running at.")
```
Comment (andraxylia) [inline cmd/istio_ca/main.go]: We fix one bug and we introduce UX limitation. With the current code and yaml templates, users can do a sed '/istio-system/kube-system' and still have a cluster wide installation. With this change, they will have to also start the CA with the flag --ca-namespace "kube-system". It is an acceptable trade-off as long as this gets documented. Even better would be to use POD_NAMESPACE for this.

Please replace "Specify whether the current CA is running at." with "Namespace where the Istio CA pods is running. If not specified, it defaults to POD_NAMESPACE".

In 0.3, we will need to swap the flags, since all controller use "namespace" mapped to POD_NAMESPACE, to indicate where the controller is running, and other flag names to indicate where they listen to. This will be a backward incompatible change, thus my recommendation was to do it properly in 0.2 , if we must fix the critical bug.

Comment (wattli) [reply cmd/istio_ca/main.go]: Thanks Andra for your input.

Comment of ca-namespace flag is updated.

Currently it's NAMESPACE, it's just one line change for me to use POD_NAMESPACE, do you think it is safe to change to POD_NAMESPACE for now?

Agreed on 0.3 vision.

File: cmd/istio_ca/main.go (hunk 17 above)
Comment (wattli) [reply cmd/istio_ca/main.go]: It's reverted.

File: cmd/istio_ca/main.go (hunk 19 above)
Comment (costinm) [reply cmd/istio_ca/main.go]: How about "Specify the namespace where Istio CA root CA is stored. Will not be used if explicit file or other storage mechanism is specified"

And rename it to 'istio-ca-storage-namespace' - to make it very clear it is specific to the CA storage (and avoid mixing it up with what Andra is concerned about )

File: pkg/pki/ca/ca.go (hunk 4 above)
Comment (costinm) [reply pkg/pki/ca/ca.go]: IMHO it's probably ok to directly use it - in future we can refactor when other storage mechanisms are introduced, but too early to abstract it.

File: pkg/pki/ca/ca.go [hunk 20]
```diff
@@ -22,10 +22,24 @@ import (
...
+	"github.com/golang/glog"
 	"istio.io/auth/pkg/pki"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	apiv1 "k8s.io/client-go/pkg/api/v1"
 )
 
 const (
+	// istioSecretType is the Istio secret annotation type.
+	istioSecretType = "istio.io/key-and-cert"
```
Comment (costinm) [inline pkg/pki/ca/ca.go]: 'key-and-cert- is too generic - maybe 'ca-root' ?

File: pkg/pki/ca/ca.go [hunk 21]
```diff
@@ -56,24 +70,57 @@ type IstioCA struct {
...
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      cASecret,
+				Namespace: namespace,
+			},
+			Type: istioSecretType,
+		}
+		_, err := core.Secrets(namespace).Create(secret)
+		if err != nil {
+			return nil, err
```
Comment (costinm) [inline pkg/pki/ca/ca.go]: IMHO - log an error ("can't save the cert") - but keep running. Even if the CA restart, we will eventually recover - however if CA fails to start 
pretty much everything will stop working.

File: cmd/istio_ca/main.go (hunk 19 above)
Comment (wattli) [reply cmd/istio_ca/main.go]: Comments and flag name are both updated

File: pkg/pki/ca/ca.go (hunk 21 above)
Comment (wattli) [reply pkg/pki/ca/ca.go]: Good point, done.

File: pkg/pki/ca/ca.go (hunk 20 above)
Comment (wattli) [reply pkg/pki/ca/ca.go]: done.

**Comment by wattli** [comment]
Shriram approved it as well (somehow dismissed by one of my commit).
